src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -108,30 +129,131 @@ impl HummockSnapshotManager {
+                }
+                Some(local_snapshot_manager) => local_snapshot_manager,
+            };
+
+            let last_epoch = local_snapshot_manager.core.lock().await.last_pinned;
+
+            match Self::pin_epoch_with_retry(
+                local_snapshot_manager.meta_client.clone(),
+                last_epoch,
+                usize::MAX,
+                || {
+                    // Should stop when the `local_version_manager` in this thread is the only
+                    // strong reference to the object.
+                    local_snapshot_manager_weak.strong_count() == 1
+                },
+            )
+            .await
+            {
+                Some(Ok(pinned_epoch)) => {
+                    let mut core_guard = local_snapshot_manager.core.lock().await;
+                    if core_guard.last_pinned < pinned_epoch {
+                        tracing::info!("Pin epoch {:?}", pinned_epoch);
+                        core_guard.last_pinned = pinned_epoch;
+                        core_guard
+                            .epoch_to_query_ids
+                            .insert(pinned_epoch, HashSet::new());
+                    }
+                }
+                Some(Err(_)) => {
+                    unreachable!(
+                        "since the max_retry is `usize::MAX`, this should never return `Err`"
+                    );
+                }
+                None => {
+                    tracing::info!("Shutdown snapshot pin worker");
+                    return;
+                }
+            };
+        }
+    }
+
     /// Used in `ObserverManager`.
     pub async fn update_snapshot_status(&self, epoch: u64) {
         let mut core_guard = self.core.lock().await;
         if core_guard.last_pinned < epoch {
-            core_guard.is_outdated = true;
+            let epoch = Self::pin_epoch_with_retry(
+                self.meta_client.clone(),
+                Default::default(),
+                usize::MAX,
+                || false,
+            )
+            .await
+            .expect("should be `Some` since `break_condition` is always false")
+            .expect("should be able to pinned");
+            core_guard.last_pinned = epoch;
+            core_guard.epoch_to_query_ids.insert(epoch, HashSet::new());
         }
     }
 }
 
 #[derive(Default)]
 struct HummockSnapshotManagerCore {
-    is_outdated: bool,
     last_pinned: u64,
     /// Record the query ids that pin each snapshot.
     /// Send an `unpin_snapshot` RPC when a snapshot is not pinned any more.
     epoch_to_query_ids: BTreeMap<u64, HashSet<QueryId>>,
 }
 
 impl HummockSnapshotManagerCore {
-    fn new() -> Self {
+    fn new(last_pinned: u64) -> Self {
         Self {
-            // Initialize by setting `is_outdated` to `true`.
-            is_outdated: true,
-            ..Default::default()
+            last_pinned,
+            epoch_to_query_ids: BTreeMap::from([(last_pinned, HashSet::new())]),
```

Comment:
Actually the last_pinned can be removed, there is retryable problems in quip: https://singularity-data.quip.com/Qbv8AypJHU68/Retryable-pin-RPC-in-hummock after we introduce unpin_snapshot_before